pkg/server/plugin/nodeattestor/tpmdevid/devid.go

```diff
@@ -0,0 +1,491 @@
+	err = verifyEKSignature(ekCert, ekRoots)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Unauthenticated, "cannot verify EK signature: %v", err)
+	}
+
+	// Verify DevID resides in the same TPM than AK
+	err = verifyDevIDCertification(&akPub, &devIDPub, attData.CertifiedDevID, attData.CertificationSignature)
+	if err != nil {
+		return nil, nil, status.Errorf(codes.Unauthenticated, "cannot to verify that DevID is in the same TPM than AK: %v", err)
+	}
+
+	// Issue a credential activation challenge (to verify AK is in the same TPM than EK)
+	challenge, nonce, err := newCredActivationChallenge(akPub, ekPub)
+	if err != nil {
+		return nil, nil, status.Error(codes.Internal, "cannot generate credential activation challenge")
+	}
+
+	return challenge, nonce, nil
+}
+
+func isDevIDResidencyInfoComplete(attReq *common_devid.AttestationRequest) error {
+	if len(attReq.AKPub) == 0 {
+		return status.Error(codes.InvalidArgument, "missing attestation public key")
+	}
+
+	if len(attReq.DevIDPub) == 0 {
+		return status.Error(codes.InvalidArgument, "missing DevID public key")
+	}
+
+	if len(attReq.EKCert) == 0 {
+		return status.Error(codes.InvalidArgument, "missing endorsement certificate")
+	}
+
+	if len(attReq.EKPub) == 0 {
+		return status.Error(codes.InvalidArgument, "missing endorsement public key")
+	}
+
+	return nil
+}
+
+func verifyEKSignature(ekCert *x509.Certificate, roots *x509.CertPool) error {
+	// Check UnhandledCriticalExtensions for OIDs that we know what to do about
+	// it (e.g. it's safe to ignore)
+	subjectAlternativeNameOID := asn1.ObjectIdentifier{2, 5, 29, 17}
+	unhandledExtensions := []asn1.ObjectIdentifier{}
+	for _, oid := range ekCert.UnhandledCriticalExtensions {
+		// Endorsement certificate's SAN is not fully processed by x509 package
+		if !oid.Equal(subjectAlternativeNameOID) {
+			unhandledExtensions = append(unhandledExtensions, oid)
+		}
+	}
+
+	ekCert.UnhandledCriticalExtensions = unhandledExtensions
+
+	_, err := ekCert.Verify(x509.VerifyOptions{
+		KeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+		Roots:     roots,
+	})
+	if err != nil {
+		return fmt.Errorf("endorsement certificate verification failed: %w", err)
+	}
+
+	return nil
+}
+
+// verifyEKsMatch checks that the public key generated using the EK template
+// matches the public key included in the Endorsement Certificate.
+func verifyEKsMatch(ekCert *x509.Certificate, ekPub tpm2.Public) error {
+	keyFromCert, ok := ekCert.PublicKey.(*rsa.PublicKey)
+	if !ok {
+		return fmt.Errorf("key from certificate is not an RSA key")
+	}
+
+	cryptoKey, err := ekPub.Key()
+	if err != nil {
+		return fmt.Errorf("cannot get template key: %w", err)
+	}
+
+	keyFromTemplate, ok := cryptoKey.(*rsa.PublicKey)
+	if !ok {
+		return fmt.Errorf("key from template is not an RSA key")
+	}
+
+	if keyFromCert.E != keyFromTemplate.E {
+		return fmt.Errorf("exponent missmatch")
+	}
+
+	if keyFromCert.N.Cmp(keyFromTemplate.N) != 0 {
+		return fmt.Errorf("modulus missmatch")
```

Comment:
No need for formatted error.
There is also a typo:
```suggestion
		return errors.New("modulus mismatch")
```